src/internal/ts.rs

```diff
@@ -38,31 +37,3 @@ impl TransactionSet {
         self.0.get_mut()
     }
 }
-
-/// Crate-public wrapper for acquiring and releasing the global transaction set
-/// which also cleans it prior to unlocking it.
-pub(crate) struct GlobalTS(MutexGuard<'static, GlobalState>);
```

Comment:
I am trying to remember what the [(now deleted) comments about RPM4 thread safety were based on my discussions with an RPM developer](https://github.com/rpm-software-management/rpm/issues/429#issuecomment-400461227), but this was here rather deliberately as my understanding is RPM4 is not thread safe whatsoever (apparently they started adding thread safety in RPM5), so I don't think it's wise to remove this as it appears this change now allows several threads to open `TransactionSet`s at once.

Reply:
Here is the only bit of @n3npq's original comment about this I see quoted:

> you can take out a static global transaction which can be used where needed w/o a need to pass an rpmts through all the methods

Re #12, I'm wondering if `rpmtsClean()` isn't being called?

Reply:
What I found when researching whether these DB locks are internal or external, is that when beginning a transaction they're locking their database. Makes sense. So we should lock up on their side in case multiple threads are going to access the database at the same time.

Looking at it again now, though, I see you're right. There are sources for `rpmtsInitIterator`:

```c

/* keyp might no be defined. */
rpmdbMatchIterator rpmtsInitIterator(const rpmts ts, rpmDbiTagVal rpmtag,
			const void * keyp, size_t keylen)
{
    rpmdbMatchIterator mi = NULL;
    char *tmp = NULL;

    if (ts == NULL)
	return NULL;

    if (ts && ts->keyring == NULL)
	loadKeyring(ts);

    if (ts->rdb == NULL && rpmtsOpenDB(ts, ts->dbmode))
	return NULL;
```

and `rpmtsOpenDB`:

```
int rpmtsOpenDB(rpmts ts, int dbmode)
{
    int rc = 0;

    if (ts->rdb != NULL && ts->dbmode == dbmode)
	return 0;

    (void) rpmtsCloseDB(ts);

    /* XXX there's a potential db lock race here. */

    ts->dbmode = dbmode;
    rc = rpmdbOpen(ts->rootDir, &ts->rdb, ts->dbmode, 0644);
```

Reply:
I discovered an issue around `failchk` used by Berkeley DB inside of RPM.

```
[root@de10012ed987 work]# cargo test --no-default-features --target-dir docker-target
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running docker-target/debug/deps/librpm-c0c0eeeaf7290b3b

running 1 test
test db::tests::db_opens ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running docker-target/debug/deps/new_api-fc2aa692809dee67

running 1 test
test db_open_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running docker-target/debug/deps/new_api_double_open-35ae8c8a7f36f7c5

running 1 test
test db_open_twice_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running docker-target/debug/deps/new_api_find-bda333c4287dfc06

running 1 test
test db_find_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running docker-target/debug/deps/new_api_find_multiple-d171c3efaa849a66

running 1 test
test db_find_test_multiple ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running docker-target/debug/deps/new_api_find_multiple_threads-6e90d0b60254ad84

running 1 test
test db_find_test_multiple ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

     Running docker-target/debug/deps/new_api_open_with-8b40f8a6640aec6b

running 1 test
test db_open_test ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests librpm

running 1 test
test src/lib.rs - db (line 37) ... FAILED

failures:

---- src/lib.rs - db (line 37) stdout ----
Test executable failed (exit code 101).

stderr:
error: rpmdb: BDB0113 Thread/process 187/139997774956288 failed: BDB1507 Thread died in Berkeley DB library
error: db5 error(-30973) from dbenv->failchk: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
error: cannot open Packages index using db5 -  (-30973)
error: cannot open Packages database in 
error: rpmdb: BDB0113 Thread/process 187/139997774956288 failed: BDB1507 Thread died in Berkeley DB library
error: db5 error(-30973) from dbenv->failchk: BDB0087 DB_RUNRECOVERY: Fatal error, run database recovery
error: cannot open Packages database in 
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:378:21
stack backtrace:
   0:     0x559c0aa9d814 - backtrace::backtrace::libunwind::trace::hda41dbcdfba36aa0
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/libunwind.rs:88
   1:     0x559c0aa9d814 - backtrace::backtrace::trace_unsynchronized::h1a8d6e1f8cb3f5d4
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/mod.rs:66
   2:     0x559c0aa9d814 - std::sys_common::backtrace::_print_fmt::h610c4127487e10da
                               at src/libstd/sys_common/backtrace.rs:76
   3:     0x559c0aa9d814 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h0722dc552e01bd1d
                               at src/libstd/sys_common/backtrace.rs:60
   4:     0x559c0aab69cc - core::fmt::write::h01edf6dd68a42c9c
                               at src/libcore/fmt/mod.rs:1030
   5:     0x559c0aa9bd17 - std::io::Write::write_fmt::hf15985f193f03c04
                               at src/libstd/io/mod.rs:1412
   6:     0x559c0aa9fb85 - std::sys_common::backtrace::_print::hd8d5d08a1795e743
                               at src/libstd/sys_common/backtrace.rs:64
   7:     0x559c0aa9fb85 - std::sys_common::backtrace::print::hf89a79e3921a2366
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x559c0aa9fb85 - std::panicking::default_hook::{{closure}}::h3a8f42beb3bb8ae3
                               at src/libstd/panicking.rs:196
   9:     0x559c0aa9f876 - std::panicking::default_hook::h8f803b0bc31a5c37
                               at src/libstd/panicking.rs:210
  10:     0x559c0aaa01d5 - std::panicking::rust_panic_with_hook::h825f041245da8739
                               at src/libstd/panicking.rs:473
  11:     0x559c0aa9fd72 - std::panicking::continue_panic_fmt::hbe0378e33481e81b
                               at src/libstd/panicking.rs:380
  12:     0x559c0aa9fc66 - rust_begin_unwind
                               at src/libstd/panicking.rs:307
  13:     0x559c0aab42ba - core::panicking::panic_fmt::h527855ce0bc891f6
                               at src/libcore/panicking.rs:85
  14:     0x559c0aab41f9 - core::panicking::panic::h2f49f09cf859b728
                               at src/libcore/panicking.rs:49
  15:     0x559c0aa8a9ce - core::option::Option<T>::unwrap::h8c5d7730af509e3d
  16:     0x559c0aa8bf48 - rust_out::main::h1cf9155cb6204ee7
  17:     0x559c0aa8a353 - std::rt::lang_start::{{closure}}::h818a47d864931e56
  18:     0x559c0aa9fc53 - std::rt::lang_start_internal::{{closure}}::ha04574f12d97cbe2
                               at src/libstd/rt.rs:49
  19:     0x559c0aa9fc53 - std::panicking::try::do_call::h7c2a8488f72db90c
                               at src/libstd/panicking.rs:292
  20:     0x559c0aaa167a - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:80
  21:     0x559c0aaa06cd - std::panicking::try::hc3a9b5da4250385d
                               at src/libstd/panicking.rs:271
  22:     0x559c0aaa06cd - std::panic::catch_unwind::hf27600bf8c37809a
                               at src/libstd/panic.rs:394
  23:     0x559c0aaa06cd - std::rt::lang_start_internal::h409d4f2fe51133b0
                               at src/libstd/rt.rs:48
  24:     0x559c0aa8a338 - std::rt::lang_start::h1e446caf94a57676
  25:     0x559c0aa8c16b - main
  26:     0x7fabc0fa2505 - __libc_start_main
                               at ../csu/libc-start.c:266
  27:     0x559c0aa8a1e9 - <unknown>
  28:                0x0 - <unknown>



failures:
    src/lib.rs - db (line 37)

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--doc'
```

However I wasn't able to reproduce it when running CentOS not in Docker but in a virtual machine. I tried debugging it in docker but didn't reach a conclusion and I suspect it's impossible w/o help from RPM developers.

This looks like an issue with `failchk` logic rather than actual problem, because there's no real multithreaded access, and simple `rpm --rebuilddb` fixes the problem. I don't even have to remove lock files or anything.

Reply:
do not use bdb, use some other backend :)

Reply:
@ignatenkobrain how would I do that and keep compatibility with actual RPM databases used in distributives like CentOS?